### PR TITLE
Add IResult implementations for more IActionResults

### PR DESCRIPTION
**Pull request by uabarahona**
# Summary 
These changes add necessary implementation to IActionResults that does not have it

# Details
The IActionResults that are going to be worked on this PR are:

- [x] RedirectResult
- [x] VirtualFileResult
- [x] FileContentResult
- [x] PhysicalFileResult

Task that needs to be done for this PR are:

- [x] Basic implementatio for IActionResults
- [x] Add test suite to test the new IActionResults

Addresses #32565

### CODE REVIEW

**Comment by barahonajm** [comment]
@halter73 @davidfowl I would like to confirm the approach I am taking is how you were envisioning the implementation of these new `IActionResults`, could you take a look at #32647?

If the approach its fine I will continue with the rest but there is a question, similar to MVC we will need to register singleton of each ActionResult like this

https://github.com/dotnet/aspnetcore/blob/d8906c8523f071371ce95d4e2d2fdfa89858047e/src/Mvc/Mvc.Core/src/DependencyInjection/MvcCoreServiceCollectionExtensions.cs#L245-L259

Where do you think is the best place to add these new addings in this particular case?

File: src/Http/Http.Abstractions/src/IResultExecutor.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    /// </remarks>
+    public interface IResultExecutor<in TResult> where TResult : notnull, IResult
+    {
+        /// <summary>
+        /// Asynchronously executes the action result, by modifying the <see cref="HttpResponse"/>.
+        /// </summary>
+        /// <param name="context">The <see cref="HttpContext"/> associated with the current request."/></param>
+        /// <param name="result">The action result to execute.</param>
+        /// <returns>A <see cref="Task"/> which represents the asynchronous operation.</returns>
+        Task ExecuteAsync(HttpContext context, TResult result);
```
Comment (halter73) [inline src/Http/Http.Abstractions/src/IResultExecutor.cs:23]: Thanks for submitting a PR. While having a parallel `IResultExecutor<TResult>` is an interesting idea, this goes a little above-and-beyond what I was thinking about for #32565.

I would look to JsonResult as an example:

https://github.com/dotnet/aspnetcore/blob/302246df4194a779fd526818a09976b7ed739d18/src/Mvc/Mvc.Core/src/JsonResult.cs#L72-L94

It bypasses the `IActionResultExecutor<JsonResult>` and does the default thing when called via `IResult.ExecuteAsync`.

File: src/Mvc/Mvc.Core/src/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -544,6 +545,8 @@ Microsoft.AspNetCore.Mvc.Infrastructure.ActionDescriptorCollection.Items.get ->
 Microsoft.AspNetCore.Mvc.Infrastructure.AmbiguousActionException.AmbiguousActionException(System.Runtime.Serialization.SerializationInfo! info, System.Runtime.Serialization.StreamingContext context) -> void
 Microsoft.AspNetCore.Mvc.Infrastructure.AmbiguousActionException.AmbiguousActionException(string? message) -> void
 Microsoft.AspNetCore.Mvc.Infrastructure.ContentResultExecutor.ContentResultExecutor(Microsoft.Extensions.Logging.ILogger<Microsoft.AspNetCore.Mvc.Infrastructure.ContentResultExecutor!>! logger, Microsoft.AspNetCore.Mvc.Infrastructure.IHttpResponseStreamWriterFactory! httpResponseStreamWriterFactory) -> void
+static Microsoft.AspNetCore.Mvc.Core.Routing.UrlHelpers.Content(Microsoft.AspNetCore.Http.HttpContext! httpContext, string? contentPath) -> string?
+static Microsoft.AspNetCore.Mvc.Core.Routing.UrlHelpers.IsLocalUrl(string? url) -> bool
```
Comment (halter73) [inline src/Mvc/Mvc.Core/src/PublicAPI.Unshipped.txt:549]: Let's try to keep the public API changes to a minimum. If you think we should add these, I suggest submitting an [API proposal](https://github.com/dotnet/aspnetcore/issues/new/choose).

File: src/Http/Http.Abstractions/src/IResultExecutor.cs (hunk 1 above)
Comment (barahonajm) [reply src/Http/Http.Abstractions/src/IResultExecutor.cs:23]: Got it, so I'm probably wrong but for this case `RedirectResult` I believe it should have almost the same implementation as:

https://github.com/dotnet/aspnetcore/blob/3d0f22b8b9df34c6b53976563fc12c6267118c75/src/Mvc/Mvc.Core/src/Infrastructure/RedirectResultExecutor.cs#L44-L78

is that correct?

File: src/Mvc/Mvc.Core/src/PublicAPI.Unshipped.txt (hunk 2 above)
Comment (barahonajm) [reply src/Mvc/Mvc.Core/src/PublicAPI.Unshipped.txt:549]: Sure, based on the answer to the previos comment I might not need any of these. thanks!

File: src/Http/Http.Abstractions/src/IResultExecutor.cs (hunk 1 above)
Comment (barahonajm) [reply src/Http/Http.Abstractions/src/IResultExecutor.cs:23]: Thanks, I got my response from #32679, I feel we are repeating code here, but I am also aware the slight differences may will make a bit hard to reuse existing code so I will follow the pattern done on #32679 and bypass other implementations.

**Comment by barahonajm** [comment]
@halter73 I have added implementation for:

-  RedirectResult
-  VirtualFileResult
-  FileContentResult

I will work on others tomorrow either on this PR or in a new one if this get accepted

File: src/Mvc/Mvc.Core/test/FileContentResultTest.cs [hunk 3]
```diff
@@ -71,8 +72,11 @@ public void Constructor_SetsLastModifiedAndEtag()
             MediaTypeAssert.Equal(expectedMediaType, result.ContentType);
         }
 
-        [Fact]
-        public async Task WriteFileAsync_CopiesBuffer_ToOutputStream()
+        [Theory]
+        [MemberData(nameof(GetActions))]
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/test/FileContentResultTest.cs:76]: I know this sounds onerous, but could we split these apart as separate tests? We don't want TheoryData as a way to drive different test code paths. It makes it harder to diagnose, skip individual scenarios that have bugs etc.

File: src/Mvc/Mvc.Core/test/PhysicalFileResultTest.cs [hunk 4]
```diff
@@ -90,8 +90,11 @@ public async Task WriteFileAsync_WritesRangeRequested(long? start, long? end, st
             Assert.Equal((long?)contentLength, sendFile.Length);
         }
 
-        [Fact]
-        public async Task WriteFileAsync_IfRangeHeaderValid_WritesRequestedRange()
+        [Theory]
+        [MemberData(nameof(GetActions))]
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/test/PhysicalFileResultTest.cs:94]: Same feedback as the other one.

**Review by pranavkm — COMMENTED** [review]
Looks really good!

File: src/Mvc/Mvc.Core/test/FileContentResultTest.cs (hunk 3 above)
Comment (barahonajm) [reply src/Mvc/Mvc.Core/test/FileContentResultTest.cs:76]: Sure, that make sense, I will add a new TestFile and duplicate de tests calling IResult.ExecuteAsync instead, thanks!

Comment (davidfowl) [reply src/Mvc/Mvc.Core/test/FileContentResultTest.cs:76]: Can we share tests with a base class or is that not possible?

Comment (barahonajm) [reply src/Mvc/Mvc.Core/test/FileContentResultTest.cs:76]: Sure, I was planning to create base clases but for share the infrastructure for the test than to share the test itself, I am going to explore that idea, I can possibly have the methods extracted in a base clase and call those in the implementations.

Comment (barahonajm) [reply src/Mvc/Mvc.Core/test/FileContentResultTest.cs]: @pranavkm @davidfowl I have used some static methods to share the implementation of the test but the tests itself are separated and in different files, let me know if that is acceptable or if I should simple rollback to separate files and duplicate test. thanks!

File: src/Mvc/Mvc.Core/test/BaseRedirectResultTest.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        private static ActionContext GetActionContext(HttpContext httpContext)
+        {
+            var routeData = new RouteData();
+            routeData.Routers.Add(new Mock<IRouter>().Object);
+
+            return new ActionContext(httpContext,
+                                    routeData,
+                                    new ActionDescriptor());
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/test/BaseRedirectResultTest.cs:69]:
```suggestion
            return new ActionContext(
                httpContext,
                routeData,
                new ActionDescriptor());
```

File: src/Mvc/Mvc.Core/test/BaseRedirectResultTest.cs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            // Assert
+            // Verifying if Redirect was called with the specific Url and parameter flag.
+            Assert.Equal(expectedPath, httpContext.Response.Headers.Location.ToString());
+            Assert.Equal(StatusCodes.Status302Found, httpContext.Response.StatusCode);
+        }
+
+        private static ActionContext GetActionContext(HttpContext httpContext)
+        {
+            var routeData = new RouteData();
+            routeData.Routers.Add(new Mock<IRouter>().Object);
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/test/BaseRedirectResultTest.cs:65]:
```suggestion
            routeData.Routers.Add(Mock.Of<IRouter>());
```

File: src/Mvc/Mvc.Core/test/BaseRedirectResultTest.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            // Act
+            await function(result, action == "ActionContext" ? actionContext : httpContext);
+
+            // Assert
+            // Verifying if Redirect was called with the specific Url and parameter flag.
+            Assert.Equal(expectedPath, httpContext.Response.Headers.Location.ToString());
+            Assert.Equal(StatusCodes.Status302Found, httpContext.Response.StatusCode);
+        }
+
+        public static async Task Execute_ReturnsAppRelativePath_WhenItStartsWithTilde(
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/test/BaseRedirectResultTest.cs:41]: If you're super up to it, you might be able to use a generic type parameter to make this work:

```suggestion
        public static async Task Execute_ReturnsAppRelativePath_WhenItStartsWithTilde<TContext>(
           ... 
           Func<RedirectResult, TContext, Task> function)
        {
        
           ...
           await function(result, typeof(TContext) == typeof(HttpContext) ? httpContext : actionContext);
```

**Review by pranavkm — APPROVED** [review]

File: src/Mvc/Mvc.Core/test/BaseRedirectResultTest.cs (hunk 7 above)
Comment (barahonajm) [reply src/Mvc/Mvc.Core/test/BaseRedirectResultTest.cs:41]: Make sense, that string was bothering me a bit :D, will use this. Thanks!

**Comment by pranavkm** [comment]
Thanks for the PR!

**Pull request merged**